pep-0xxx.rst

```diff
@@ -0,0 +1,1149 @@
+                lowest_supported_version = self.lowest_supported_version
+
+            if highest_supported_version is _DEFAULT_VALUE:
+                highest_supported_version = self.highest_supported_version
+
+            if trust_store is _DEFAULT_VALUE:
+                trust_store = self.trust_store
+
+            if sni_callback is _DEFAULT_VALUE:
+                sni_callback = self.sni_callback
+
+            return self.__class__(
+                validate_certificates, certificate_chain, ciphers,
+                inner_protocols, lowest_supported_version,
+                highest_supported_version, trust_store, sni_callback
+            )
+
+
+
+Context
+~~~~~~~
+
+The ``Context`` abstract base class defines an object that allows configuration
+of TLS. It can be thought of as a factory for ``TLSWrappedSocket`` and
+``TLSWrappedBuffer`` objects.
+
+As much as possible implementers should aim to make these classes immutable:
+that is, they should prefer not to allow users to mutate their internal state
+directly, instead preferring to create new contexts from new TLSConfiguration
+objects. Obviously, the ABCs cannot enforce this constraint, and so they do not
+attempt to.
+
+The ``Context`` abstract base class has the following class definition::
+
+    TLSBufferObject = Union[TLSWrappedSocket, TLSWrappedBuffer]
+
+
+    class _BaseContext(metaclass=ABCMeta):
+        @abstractmethod
+        def __init__(self, configuration: TLSConfiguration):
+            """
+            Create a new context object from a given TLS configuration.
+            """
+
+        @property
+        @abstractmethod
+        def configuration(self) -> TLSConfiguration:
+            """
+            Returns the TLS configuration that was used to create the context.
+            """
+
+
+    class ClientContext(_BaseContext):
+        @abstractmethod
+        def wrap_socket(self, socket: socket.socket,
+                        auto_handshake=True: bool,
+                        server_hostname=None: Optional[str]) -> TLSWrappedSocket:
+            """
+            Wrap an existing Python socket object ``socket`` and return a
+            ``TLSWrappedSocket`` object. ``socket`` must be a ``SOCK_STREAM``
+            socket: all other socket types are unsupported.
+
+            The returned SSL socket is tied to the context, its settings and
+            certificates.
+
+            The parameter ``auto_handshake`` specifies whether to do the SSL
+            handshake automatically after doing a ``socket.connect()``, or
+            whether the application program will call it explicitly, by
+            invoking the ``TLSWrappedSocket.do_handshake()`` method. Calling
+            ``TLSWrappedSocket.do_handshake()`` explicitly gives the program
+            control over the blocking behavior of the socket I/O involved in
+            the handshake.
+
+            The optional parameter ``server_hostname`` specifies the hostname
```

Comment:
I'm with Donald, one parameter is easier to handle. For Python 3.7 I had this idea:

* ditch ```ssl.match_hostname()``` and use OpenSSL 1.0.2's verification features
* have some sort of additional parameter to change settings (wildcard policy, CN match, IP address validation on/off)
* use ```inet_pton``` to detect if the ```server_hostname``` string is a IPv4 or IPv6 address. The function is pretty fast and efficient.
* in case ```server_hostname``` is a host name, send it as SNI automatically
* in case hostname verification is enabled and ```server_hostname``` is a host name, use ```server_hostname``` to verify with SAN DNS name
* in case IP address verification is enabled and it is an IP address, use ```server_hostname``` to check with SAN IPAddress

In the future the schema can be enhanced for SRV-ID:
* if server_hostname is a hostname that starts with a ```_```, strip the service indicator, use the rest as SNI and perform SRV-ID matching with service indicator and hostname.

I also had the idea to use ```ipaddress.IPAddress``` for IP address matching. I'm not sure how well the ```ipaddress``` module is supported in ```socket```.